src/mem/x86_64.rs

```diff
@@ -178,21 +178,78 @@ pub unsafe fn c_string_length(s: *const core::ffi::c_char) -> usize {
     let mut n: usize;
 
     asm!(
-        // search for a zero byte
+        // For small sizes, we avoid invoking SSE instructions.
+        // make manual comparisons instead.
         "xor %eax, %eax",
+        "cmpb $0, (%rdi)",
+        "je 3f",
+        "mov $1, %eax",
+        "cmpb $0, 1(%rdi)",
+        "je 3f",
+        "mov $2, %eax",
+        "cmpb $0, 2(%rdi)",
+        "je 3f",
+        "mov $3, %eax",
+        "cmpb $0, 3(%rdi)",
+        "je 3f",
 
-        // unbounded memory region
-        "xor %ecx, %ecx",
-        "not %rcx",
+        // Adjust address
+        "add $4, %rdi",
 
-        // perform search
-        "repne scasb (%rdi), %al",
+        // Align the address to 16 bytes (xmm register size).
+        // This is important, since an n byte read
+        // with n byte alignment is guranteed to never cross
+        // a page boundary and thus will never try to access
+        // memory which may not be accessible.
+        "mov %edi, %ecx",
+        "and $15, %ecx",
+        "and $-16, %rdi",
 
-        // extract length
-        "not %rcx",
-        "dec %rcx",
+        // zero out an xmm register for comparisons with zero.
+        "pxor    %xmm0, %xmm0",
```

Comment:
Are you using `#[cfg(target_feature="sse2")]`? Kernels generally have SSE and SSE2 disabled to avoid having to save and restore vector regs on syscalls.

Reply:
I was only thinking about userspace scenarios. I fixed this.